### DISCUSSION TITLE
Vector Tiles in EPSG:2056

### DISCUSSION THREAD

**Discussion by FabianRechsteiner** [developmentseed/tipg#165] [Q&A]
I would like to make my geospatial data available as Vector Tiles in the national coordinate system of Switzerland, EPSG:2056. Is it possible to support this as well?
Example: [Geoserver](https://ogc.heig-vd.ch/geoserver/ogc/tiles/tileMatrixSets/EPSG%3A2056)

**Comment by vincentsarago** [comment]
Hi @FabianRechsteiner 
This is absolutely doable. 

In `tipg` we rely on [morecantile](https://github.com/developmentseed/morecantile) for the TileMatrixSet definition. The list of supported TMS is then passed as an option when we create the endpoints https://github.com/developmentseed/tipg/blob/80a86a999cfbd52556476cda20400b5a1c3e8fce/tipg/factory.py#L1845

You have multiple choice for customization, but first you'll need to create a TileMatrixSet JSON document following the specs found in the geoserver docs (https://ogc.heig-vd.ch/geoserver/ogc/tiles/tileMatrixSets/EPSG%3A2056?f=application%2Fjson should work) Then you need to tell the endpoints that your TMS is available:

1. use `TILEMATRIXSET_DIRECTORY` env variable (https://developmentseed.org/morecantile/usage/#automatically-register-tms-documents) 

If you place you json file in a directory and then set `TILEMATRIXSET_DIRECTORY` environment, Morecantile will list the TMS automagically.

or

2. Create a custom application

This will require more work but also enable more customization 

```python
from contextlib import asynccontextmanager
from typing import Any, List

import jinja2

from tipg.collections import register_collection_catalog
from tipg.database import close_db_connection, connect_to_db
from tipg.errors import DEFAULT_STATUS_CODES, add_exception_handlers
from tipg.factory import Endpoints
from tipg.middleware import CacheControlMiddleware, CatalogUpdateMiddleware
from tipg.settings import (
    APISettings,
    CustomSQLSettings,
    DatabaseSettings,
    PostgresSettings,
)

from fastapi import FastAPI, Request

from starlette.middleware.cors import CORSMiddleware
from starlette.templating import Jinja2Templates
from starlette_cramjam.middleware import CompressionMiddleware

import morecantile

settings = APISettings()
postgres_settings = PostgresSettings()
db_settings = DatabaseSettings()
custom_sql_settings = CustomSQLSettings()

default_tms = morecantile.tms.register({"MyCustomTmsEPSG2056": "path to the TMS file"})

@asynccontextmanager
async def lifespan(app: FastAPI):
    """FastAPI Lifespan."""
    # Create Connection Pool
    await connect_to_db(
        app,
        settings=postgres_settings,
        schemas=db_settings.schemas,
        user_sql_files=custom_sql_settings.sql_files,
    )

    # Register Collection Catalog
    await register_collection_catalog(
        app,
        schemas=db_settings.schemas,
        tables=db_settings.tables,
        exclude_tables=db_settings.exclude_tables,
        exclude_table_schemas=db_settings.exclude_table_schemas,
        functions=db_settings.functions,
        exclude_functions=db_settings.exclude_functions,
        exclude_function_schemas=db_settings.exclude_function_schemas,
        spatial=db_settings.only_spatial_tables,
        spatial_extent=db_settings.spatial_extent,
        datetime_extent=db_settings.datetime_extent,
    )

    yield

    # Close the Connection Pool
    await close_db_connection(app)


app = FastAPI(
    title=settings.name,
    openapi_url="/api",
    docs_url="/api.html",
    lifespan=lifespan,
)

ogc_api = Endpoints(
    title=settings.name,
    supported_tms=default_tms,
    with_tiles_viewer=settings.add_tiles_viewer,
)
app.include_router(ogc_api.router)

app.add_middleware(CacheControlMiddleware, cachecontrol=settings.cachecontrol)
app.add_middleware(CompressionMiddleware)

add_exception_handlers(app, DEFAULT_STATUS_CODES)
```

**Reply by FabianRechsteiner** [reply]
I tried to save the TileMatrixSet JSON from Geoserver in the directory so that it is automatically recognized as a TMS. Unfortunately, this did not work with the following settings:

I created a new directory named "tms" and saved the JSON file in it.

- tipg
  - .env
  - tms
    -  EPSG_2056.json
 
In the EPSG_2056.json file I changed the ID from `EPSG:2056 `to `EPSG_2056`.
`{"id":"EPSG_2056","...,`

In the .env file, I added the parameter `TILEMATRIXSET_DIRECTORY=tms`.

I also tried to add the Parameter in the compose.yaml File:
```
    environment:
      - POSTGRES_USER=${POSTGRES_USER_USERNAME:-user_name}
      - POSTGRES_PASS=${POSTGRES_USER_PASSWORD:-user_password}
      - POSTGRES_DBNAME=${POSTGRES_DB:-postgres}
      - POSTGRES_HOST=${POSTGRES_HOST:-postgis}
      - POSTGRES_PORT=5432
      - PORT=8080
      - DEBUG=TRUE
      - TILEMATRIXSET_DIRECTORY=tms
```
Where do i have to define the Parameter?

**Reply by vincentsarago** [reply]
the `TILEMATRIXSET_DIRECTORY=tms` seems a bit odd. I would put an absolute path just to make sure. I believe the path should be relative to where the `uvicorn` command is run.

**Reply by FabianRechsteiner** [reply]
It suddenly worked when I defined the parameter only in the compose.yaml file.

However, now the following error message appears: "Tile Matrix Set must be version 2.0. Use morecantile <4.0 for TMS 1.0 support."

So, I need to somehow migrate the JSON file to version 2.0.

Do you have any idea how to do that?

**Reply by vincentsarago** [reply]
`morecantile.TileMatrixSet` has a `from_v1` method 

```python

import json
import morecantile

with open("EPSG_2056.json", "r") as f:
   tms = json.loads(f.read())

with open("EPSG_2056.json", "w") as f:
    tms_v2 = morecantile.TileMatrixSet.from_v1(tms)
    f.write(tms_v2.model_dump_json(exclude_none=True))
```

https://github.com/developmentseed/morecantile/blob/main/morecantile/models.py#L592C9-L592C16

you can also check this PR https://github.com/developmentseed/morecantile/pull/101

**Reply by FabianRechsteiner** [reply]
Thanks for the [link](https://github.com/developmentseed/morecantile/pull/101). This showed me, that some parts of the json file is already in Version 2.0. The parameter `"type":"application/x-yaml",` is inside the json and was removed in Version 2.0. 
So i deleted the type-parameter and renamed `supportedCRS `to `crs`.

But it has still the old parameter cornerOfOrigin which cant't be changed easy.

I guess it's better, not to use this example and create a new tms instead. 🙈

Thank you anyway for your help.
```
{
    "id": "EPSG_2056",
    "links": [
        {
            "href": "https://ogc.heig-vd.ch/geoserver/ogc/tiles/tileMatrixSets/EPSG%3A2056?f=application%2Fx-yaml",
            "rel": "alternate",
            "title": "This document as application/x-yaml"
        },
        {
            "href": "https://ogc.heig-vd.ch/geoserver/ogc/tiles/tileMatrixSets/EPSG%3A2056?f=application%2Fjson",
            "rel": "self",
            "title": "This document"
        },
        {
            "href": "https://ogc.heig-vd.ch/geoserver/ogc/tiles/tileMatrixSets/EPSG%3A2056?f=text%2Fhtml",
            "rel": "alternate",
            "title": "This document as text/html"
        }
    ],
    "title": "EPSG_2056 (CH1903+ / LV95) tile matrix set.",
    "crs": "http://www.opengis.net/def/crs/EPSG/0/2056",
    "tileMatrices": [
        {
            "id": "EPSG_2056:0",
            "scaleDenominator": 1.4285714285714287E7,
            "cornerOfOrigin": "topLeft",
            "pointOfOrigin": [
                2485014.052451379,
                2098189.0
            ],
            "tileWidth": 256,
            "tileHeight": 256,
            "matrixWidth": 1,
            "matrixHeight": 1
        },
...
```

**Reply by FabianRechsteiner** [reply]
Three days ago the new TMS for switzerland was published: [Linkt to the SwissLV95CellSizes.json](https://www.ech.ch/de/ech/ech-0056/4.0.0 )


I renamed the file to SwissLV95CellSizes.json and saved it in my "tms" folder.

The TMS could be correctly interpreted.
![image](https://github.com/developmentseed/tipg/assets/43602650/5b46e400-b8a8-42fa-ae7d-7a70227e6011)

Now i tried to add my Vector-Tilese-Source in QGIS: 
http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/{z}/{x}/{y}

When I use EPSG:3857 for these vector tiles, the data is displayed in the wrong place.
![image](https://github.com/developmentseed/tipg/assets/43602650/59967729-bdcf-4c25-8c63-78ef65f065b3)
![image](https://github.com/developmentseed/tipg/assets/43602650/302869d4-5938-4828-ac57-c67d121e7644)

When I use EPSG:2056 for these vector tiles, the data is only displayed in Zoomlevel 12 and in the wrong place.
![image](https://github.com/developmentseed/tipg/assets/43602650/4c562f06-e92e-4920-8d08-3ad17f2158dc)
![image](https://github.com/developmentseed/tipg/assets/43602650/1e365972-6f24-4a13-984f-45b657c3cea3)

The TiPG Server Log showed me that:
`tipg         | 172.20.0.1:48964 - "GET /collections/avprodukt.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/3/2/5 HTTP/1.1" 200`

What ist still wrong?

Thank you for your help?

**Reply by vincentsarago** [reply]
> When I use EPSG:3857 for these vector tiles, the data is displayed in the wrong place.

to me it seems there is an issue with the original data stored in the database!

> When I use EPSG:2056 for these vector tiles, the data is only displayed in Zoomlevel 12 and in the wrong place.

must be some error messages ?

**Reply by FabianRechsteiner** [reply]
The data are correctly stored in EPSG:2056 as polygon.
There are no error messages.

The server accepts the QGIS tile requests and consistently returns empty tiles to me. This is likely because the wrong tiles are being requested or the server has stored data in the wrong tiles.

**Reply by vincentsarago** [reply]
can you store the data in epsg:4326 and check if it works?

**Reply by FabianRechsteiner** [reply]
Yes that worked.
http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_4326_mv/tiles/WebMercatorQuad/{z}/{x}/{y}

`tipg         | 172.20.0.1:54534 - "GET /collections/avprodukt.gg_gemeindegrenze_4326_mv/tiles/WebMercatorQuad/12/2145/1432 HTTP/1.1" 200`

I also tried to create a new table: avprodukt.gg_gemeindegrenze_2056_mv
```
 SELECT objectid,
    st_transform(wkb_geometry, 2056) AS wkb_geometry
   FROM avprodukt.gg_gemeindegrenze
```
Now This one has another Boundinbox: http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_2056_mv?f=json
```
"extent": {
    "spatial": {
        "bbox": [
            [
                2692010.75,
                1248144.5,
                2753917.75,
                1283679.875
            ]
        ],
        "crs": "http://www.opengis.net/def/crs/OGC/1.3/CRS84"
    }
},
```

My Table without transformation but also in EPSG:2056: avprodukt.gg_gemeindegrenze_mv
```
 SELECT objectid,
    gemeinde_name,
    gemeinde_bfsnr,
    wkb_geometry
   FROM avprodukt.gg_gemeindegrenze
WITH DATA;
```
```
"extent": {
    "spatial": {
        "bbox": [
            [
                8.661115396174294,
                47.367809019968455,
                9.4846396724923,
                47.69564287160093
            ]
        ],
        "crs": "http://www.opengis.net/def/crs/OGC/1.3/CRS84"
    }
},
```
Server Terminal:
```
tipg         | ST_Intersects: Operation on mixed SRID geometries (Polygon, 4326) != (Polygon, 2056)
tipg         | NoneType: None
tipg         | 172.20.0.1:34838 - "GET /collections/avprodukt.gg_gemeindegrenze_2056_mv/tiles/WebMercatorQuad/7/74/61 HTTP/1.1" 500
```

Transformation from 4326 to 3857 works.
Transformation from 2056 to 3857 doesn't work: http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_2056_mv/tiles/WebMercatorQuad/{z}/{x}/{y}
http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_2056_mv/tiles/Swisslv95cellsizes/{z}/{x}/{y}

There is still a problem and the data cant' be displayed.
Is there a problem in the transformation from 2056 to 3857?

**Reply by vincentsarago** [reply]
it seems to me that your original table is missing the `srid` and thus default to 4326 with the geometry coordinates beeing in 2056.

**Reply by FabianRechsteiner** [reply]
My table (matview) has a srid:
![image](https://github.com/developmentseed/tipg/assets/43602650/0c656686-9c91-4410-8e0e-a9f2f000fa4e)

**Reply by vincentsarago** [reply]
@FabianRechsteiner are you still having issue? could you share your data (one polygon will be enough) so I can try to replicate locally

**Reply by FabianRechsteiner** [reply]
Yes, I still have the same problem.

Here is my table as a GeoPackage.

[gg_gemeindegrenze_mv.zip](https://github.com/developmentseed/tipg/files/14747685/gg_gemeindegrenze_mv.zip)

**Reply by vincentsarago** [reply]
🤔 everything works well on my local machine 😬 

<img width="698" src="https://github.com/developmentseed/tipg/assets/10407788/0b665779-d313-4cfa-bd52-1605f71d9247">


I've used QGIS to load the data into postgres and made sure to keep the data in the original SRID 

<img width="398" src="https://github.com/developmentseed/tipg/assets/10407788/87c5aa67-85d0-486a-a969-b7c4db6edd15">


`http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv?f=json`

```
{
  "id": "public.gg_gemeindegrenze_mv",
  "title": "public.gg_gemeindegrenze_mv",
  "links": [
    {
      "href": "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv",
      "rel": "self",
      "type": "application/json"
    },
    {
      "href": "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv/items",
      "rel": "items",
      "type": "application/geo+json",
      "title": "Items"
    },
    {
      "href": "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv/items?f=csv",
      "rel": "alternate",
      "type": "text/csv",
      "title": "Items (CSV)"
    },
    {
      "href": "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv/items?f=geojsonseq",
      "rel": "alternate",
      "type": "application/geo+json-seq",
      "title": "Items (GeoJSONSeq)"
    },
    {
      "href": "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv/queryables",
      "rel": "queryables",
      "type": "application/schema+json",
      "title": "Queryables"
    }
  ],
  "extent": {
    "spatial": {
      "bbox": [
        [
          8.657019844813632,
          47.36615521754696,
          9.488782848219593,
          47.697267839358545
        ]
      ],
      "crs": "http://www.opengis.net/def/crs/OGC/1.3/CRS84"
    }
  },
  "itemType": "feature",
  "crs": [
    "http://www.opengis.net/def/crs/OGC/1.3/CRS84"
  ]
}
```

Could you check the `/items` endpoints, to make sure it also return proper GeoJSON in epsg4326

<img width="882" src="https://github.com/developmentseed/tipg/assets/10407788/8b19df53-6af0-4689-ac9f-c9f824cb1654">

**Reply by FabianRechsteiner** [reply]
```
{
    "id": "avprodukt.gg_gemeindegrenze_mv",
    "title": "avprodukt.gg_gemeindegrenze_mv",
    "links": [
        {
            "href": "http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv",
            "rel": "self",
            "type": "application/json"
        },
        {
            "href": "http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/items",
            "rel": "items",
            "type": "application/geo+json",
            "title": "Items"
        },
        {
            "href": "http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/items?f=csv",
            "rel": "alternate",
            "type": "text/csv",
            "title": "Items (CSV)"
        },
        {
            "href": "http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/items?f=geojsonseq",
            "rel": "alternate",
            "type": "application/geo+json-seq",
            "title": "Items (GeoJSONSeq)"
        },
        {
            "href": "http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/queryables",
            "rel": "queryables",
            "type": "application/schema+json",
            "title": "Queryables"
        }
    ],
    "extent": {
        "spatial": {
            "bbox": [
                [
                    8.661115396174294,
                    47.367809019968455,
                    9.4846396724923,
                    47.69564287160093
                ]
            ],
            "crs": "http://www.opengis.net/def/crs/OGC/1.3/CRS84"
        }
    },
    "itemType": "feature",
    "crs": [
        "http://www.opengis.net/def/crs/OGC/1.3/CRS84"
    ]
}
```

![image](https://github.com/developmentseed/tipg/assets/43602650/6afbc80a-6b41-4dea-8cd6-26affcb0ac4c)

That looks good. It seems like my data is being interpreted correctly.

Now, I have my additional TMS Swisslv95cellsizes. I want to display my data in [this ](https://github.com/developmentseed/tipg/discussions/165#discussioncomment-8783826)TMS in QGIS.

Could you please test if you can correctly display the data in QGIS with this TMS?
`http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/{z}/{x}/{y}`

**Reply by vincentsarago** [reply]
Ok I can reproduce this. Sadly I can't really tell if the issue is with tipg or QGIS. Non-WebMercatorQuad VectorTiles client are pretty rare 😬

**Reply by vincentsarago** [reply]
Notes: 

I've been able to produce Raster tiles using titiler and your custom TMS
<img width="1000" src="https://github.com/developmentseed/tipg/assets/10407788/f92143ca-4da4-4c8f-81ff-193cd481b1ee">

And then if I put both the raster and vector tiles together it seems ok 
<img width="1000" src="https://github.com/developmentseed/tipg/assets/10407788/53104834-d50e-43c1-a91f-44ea8670ee85">

But when I try the same in QGIS it doesn't work. It works with other TMS so I'm starting to wonder if the TMS is correct or if this is just a QGIS issue

If the TMS was wrong the Raster tiles won't have worked 🤷

**Reply by vincentsarago** [reply]
🤔 trying to load the Raster Tiles (XYZ) in QGIS with your TMS and it doesn't really work 😬

**Reply by FabianRechsteiner** [reply]
Here ist the official document for this TMS in German: [Page 16](https://www.ech.ch/sites/default/files/imce/eCH-Dossier/eCH-Dossier_PDF_Publikationen/Hauptdokument/STAN_d_DEF_2024-01-30_eCH-0056_V4.0.0_Anwendungsprofil%20Geodienste_0.pdf)

This TMS only has Zoom 0 - 15. Maybe this is the problem because QGIS wants to put the geometry on Zoom 1 but this is not the same zoomlevel like WebMercatorQuad. I don't know.

I have the same problem also with Maplibre.

**Reply by vincentsarago** [reply]
as far as I know MapLibre do not support other TMS than WebMercator (https://github.com/maplibre/maplibre-gl-js/issues/168) 

I usually use Leaflet 

```html
<!doctype html>
<html lang="en">
  <head>
    <style>
		  body { margin:0; padding: 0; width:100%; height:100%;}
		  .map { position:absolute; top:0; bottom:0; width:100%; }
    </style>
    <link rel="stylesheet" href="https://unpkg.com/leaflet@1.0.3/dist/leaflet.css">
    <script src="https://unpkg.com/leaflet@1.0.3/dist/leaflet.js"></script>
    <script type="text/javascript" src="https://unpkg.com/leaflet.vectorgrid@1.2.0"></script>
    <script src="https://unpkg.com/proj4leaflet@1.0.2/lib/proj4-compressed.js"></script>
    <script src="https://unpkg.com/proj4leaflet@1.0.2/src/proj4leaflet.js"></script>
  </head>
  <body>
    <div id="map" class="map"></div>

	<script type="text/javascript">
      var crs = new L.Proj.CRS(
          'http://www.opengis.net/def/crs/EPSG/0/2056',
          '+proj=somerc +lat_0=46.9524055555556 +lon_0=7.43958333333333 +k_0=1 +x_0=2600000 +y_0=1200000 +ellps=bessel +units=m +no_defs +type=crs', {
          origin: [2419995.75, 1350004.29],
          bounds: L.bounds(
            L.Point(2419995.75, 326004.29000000004),
            L.Point(3443995.75, 1350004.29)
          ),
          resolutions: [4000.0, 2000.0, 1000.0, 500.0, 250.0, 100.0, 50.0, 20.0, 10.0, 5.0, 2.5, 1.0, 0.5, 0.25, 0.1, 0.05],
        }
      );

      var map = L.map('map', {
        center: [47, 10],
        zoom: 3,
        maxZoom: 10,
        crs: crs
      });

      // L.tileLayer(
      //   "http://127.0.0.1:8080/cog/tiles/SwissLV95CellSizes/{z}/{x}/{y}@1x?url=world.tif",
      //   {
      //     minZoom: 0,
      //     maxZoom: 8,
      //     bounds: L.latLngBounds([-90, -180], [90, 180]),
      //   }
      // ).addTo(map);

      L.vectorGrid.protobuf(
        "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/{z}/{x}/{y}",
      ).addTo(map)
    </script>

  </body>
</html>
```

**Reply by FabianRechsteiner** [reply]
```html
<!DOCTYPE html>
<html>

<head>
  <title>LV95 Grid</title>
  <script src="https://cdn.jsdelivr.net/npm/ol@7.2.2/dist/ol.min.js"></script>
  <link href="https://cdn.jsdelivr.net/npm/ol@7.2.2/ol.min.css" rel="stylesheet">
  <script src="https://cdn.jsdelivr.net/npm/proj4@2.8.1/dist/proj4.min.js"></script>
  <style>
    body {
      font-family: sans, "Helvetica";
    }

    .map {
      height: 600px;
      top: 3em;
      width: 100%;
      background: #f8f4f0;
    }
  </style>
</head>

<body>
  <h2>Custom User Grid (Swiss LV95)</h2>
  <div id="map" class="map"></div>
  <script>

    /* Custom resolutions, from Swiss norm eCH-0056 (https://www.ech.ch/fr/ech/ech-0056/3.0) */
    const RESOLUTIONS = [
      4000, 3750, 3500, 3250, 3000, 2750, 2500, 2250, 2000, 1750, 1500, 1250,
      1000, 750, 650, 500, 250, 100, 50, 20, 10, 5, 2.5, 2, 1.5, 1, 0.5
    ];

    /* Registering and setting a new custom projection */
    proj4.defs("EPSG:2056", "+proj=somerc +lat_0=46.95240555555556 +lon_0=7.439583333333333 +k_0=1 +x_0=2600000 +y_0=1200000 +ellps=bessel +towgs84=674.374,15.056,405.346,0,0,0,0 +units=m +no_defs");
    ol.proj.proj4.register(proj4);

    var extent = [2420000, 1030000, 2900000, 1350000];
    var projection = ol.proj.get("EPSG:2056");
    projection.setExtent(extent);


    /* Convenience methods for trying new layers */
    const queryString = window.location.search;
    const urlParams = new URLSearchParams(queryString);
    const layername = urlParams.get('layer') || 'g1k18'
    const debug = urlParams.get('debug') || false;
    const servername = urlParams.get('server') || location.origin;

    /* Setting up the custom grid */
    var matrixIds = [];
    for (var i = 0; i < RESOLUTIONS.length; i++) {
      matrixIds.push(i);
    }

    var tileGrid = new ol.tilegrid.WMTS({
      origin: [extent[0], extent[3]],
      resolutions: RESOLUTIONS,
      matrixIds: matrixIds
    });

    /* Trusted WMTS raster layer */

    var wmtsLayer = new ol.layer.Tile({
      source: new ol.source.WMTS(({
        url: 'https://wmts.geo.admin.ch/1.0.0/{Layer}/default/current/2056/{TileMatrix}/{TileCol}/{TileRow}.jpeg',
        tileGrid: tileGrid,
        projection: projection,
        layer: "ch.swisstopo.swissimage",
        requestEncoding: 'REST'
      }))
    });

    /* Vector layer of interest  */

    var style = new ol.style.Style({
      stroke: new ol.style.Stroke({
        color: '#FFFF33',
        width: 3
      })
    });

    var vtLayer = new ol.layer.VectorTile({
      source: new ol.source.VectorTile({
        format: new ol.format.MVT(),
        tileGrid: tileGrid,
        tilePixelRatio: 16,
        url: 'http://localhost:8080/collections/avprodukt.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/{z}/{x}/{y}',
        projection: projection
      }),
      style: style
    });

    /* Displaying the tiles names, for debugging */
    const tileDebug = new ol.layer.Tile({
      source: new ol.source.TileDebug({
        projection: vtLayer.getSource().getProjection(),
        tileGrid: vtLayer.getSource().getTileGrid(),
      })
    });

    /* View and vector projection have to match, no reprojection like for raster layers */
    var map = new ol.Map({
      layers: [wmtsLayer, vtLayer],
      target: 'map',
      view: new ol.View({
        center: [2660000, 1190000],
        projection: projection,
        resolution: 500
      })
    });

    if (debug) {
      map.addLayer(tileDebug);
    }
  </script>
</body>

</html>
```

I used this map for the test.
TiPg gives me this Log:


```
tipg         |   File "/opt/bitnami/python/lib/python3.11/site-packages/tipg/collections.py", line 364, in _select_mvt
tipg         |     if not tms.is_valid(tile):
tipg         |            ^^^^^^^^^^^^^^^^^^
tipg         |   File "/opt/bitnami/python/lib/python3.11/site-packages/morecantile/models.py", line 1329, in is_valid
tipg         |     extrema = self.minmax(t.z)
tipg         |               ^^^^^^^^^^^^^^^^
tipg         |   File "/opt/bitnami/python/lib/python3.11/site-packages/morecantile/models.py", line 1316, in minmax
tipg         |     m = self.matrix(zoom)
tipg         |         ^^^^^^^^^^^^^^^^^
tipg         |   File "/opt/bitnami/python/lib/python3.11/site-packages/morecantile/models.py", line 682, in matrix
tipg         |     raise InvalidZoomError(
tipg         | morecantile.errors.InvalidZoomError: TileMatrix not found for level: 17 - Unable to construct tileMatrix for TMS with variable scale

postgis      | 2024-03-28 12:52:21.406 UTC [7896] STATEMENT:  
postgis      |              WITH
postgis      |              t AS (SELECT gemeinde_bfsnr, gemeinde_name, objectid, ST_AsMVTGeom(ST_Transform(ST_Intersection(ST_MakeEnvelope($1, $2, $3, $4, $5), ST_Transform(wkb_geometry::geometry, $6::int)), $7::int), ST_Segmentize(ST_MakeEnvelope($8, $9, $10, $11), $12), $13, $14, $15) AS geom
postgis      |  FROM avprodukt.gg_gemeindegrenze_mv
postgis      |  WHERE $16 AND ST_Intersects(ST_Transform(ST_Segmentize(ST_MakeEnvelope($17, $18, $19, $20, $21), $22), $23::int), wkb_geometry)
postgis      |  LIMIT $24)
postgis      |              SELECT ST_AsMVT(t.*, $25) FROM t
postgis      |              
postgis      | 2024-03-28 12:52:21.410 UTC [7897] ERROR:  ST_Segmentize: invalid max_distance 0 (must be >= 0)
postgis      | 2024-03-28 12:52:21.410 UTC [7897] STATEMENT:  
postgis      |              WITH
postgis      |              t AS (SELECT gemeinde_bfsnr, gemeinde_name, objectid, ST_AsMVTGeom(ST_Transform(ST_Intersection(ST_MakeEnvelope($1, $2, $3, $4, $5), ST_Transform(wkb_geometry::geometry, $6::int)), $7::int), ST_Segmentize(ST_MakeEnvelope($8, $9, $10, $11), $12), $13, $14, $15) AS geom
postgis      |  FROM avprodukt.gg_gemeindegrenze_mv
postgis      |  WHERE $16 AND ST_Intersects(ST_Transform(ST_Segmentize(ST_MakeEnvelope($17, $18, $19, $20, $21), $22), $23::int), wkb_geometry)
postgis      |  LIMIT $24)
postgis      |              SELECT ST_AsMVT(t.*, $25) FROM t
postgis      |              
tipg         | ST_Segmentize: invalid max_distance 0 (must be >= 0)
tipg         | NoneType: None
tipg         | 172.26.0.1:41632 - "GET /collections/avprodukt.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/8/-2/0 HTTP/1.1" 500

tipg         | 172.26.0.1:48480 - "GET /collections/avprodukt.gg_gemeindegrenze_mv/tiles/SwissLV95CellSizes/14/-1/-2 HTTP/1.1" 500
tipg         | /opt/bitnami/python/lib/python3.11/site-packages/morecantile/models.py:791: PointOutsideTMSBounds: Point (2420098.15, 1350055.49) is outside TMS bounds [2419995.75, 326004.29000000027, 3443995.75, 1350004.29].
tipg         |   warnings.warn(
```

**Reply by vincentsarago** [reply]
Yes I've seen this as well, this must comes from https://github.com/developmentseed/tipg/blob/1edfab5dbdcc737b475c7ab965e4c129a89870c6/tipg/collections.py#L386-L405
 
I'm not quite sure why `bbox.right - bbox.left` could be `0` or `<0`. There must be something to dig in there.

At the same time, I've saw you're also looking at Tegola, are you able to make it work with this TMS? have you also looked at other Tile server (just so we can compare and find where we have a bug)

**Reply by vincentsarago** [reply]
🤔 the JS code looks also a bit weird to me 

you have
```
    /* Custom resolutions, from Swiss norm eCH-0056 (https://www.ech.ch/fr/ech/ech-0056/3.0) */
    const RESOLUTIONS = [
      4000, 3750, 3500, 3250, 3000, 2750, 2500, 2250, 2000, 1750, 1500, 1250,
      1000, 750, 650, 500, 250, 100, 50, 20, 10, 5, 2.5, 2, 1.5, 1, 0.5
    ];
```

while looking at the TMS the resolution should be:
```
[4000.0, 2000.0, 1000.0, 500.0, 250.0, 100.0, 50.0, 20.0, 10.0, 5.0, 2.5, 1.0, 0.5, 0.25, 0.1, 0.05]
```

The CRS is also a bit different from what pyproj gives us 

```
+proj=somerc +lat_0=46.95240555555556 +lon_0=7.439583333333333 +k_0=1 +x_0=2600000 +y_0=1200000 +ellps=bessel +towgs84=674.374,15.056,405.346,0,0,0,0 +units=m +no_defs
```

```
from pyproj import CRS

CRS.from_epsg(2056).to_proj4()
>> '+proj=somerc +lat_0=46.9524055555556 +lon_0=7.43958333333333 +k_0=1 +x_0=2600000 +y_0=1200000 +ellps=bessel +units=m +no_defs +type=crs'
```

**Reply by vincentsarago** [reply]
So, I've looked at https://wmts.geo.admin.ch/EPSG/2056/1.0.0/WMTSCapabilities.xml and the TMS definition for `2056` is quite different from the `SwissLV95CellSizes` TMS and should be something like: [2056_26.json](https://github.com/developmentseed/tipg/files/14791842/2056_26.json)


When I then use this TMS with 

```html
<!doctype html>
<html lang="en">
  <head>
    <style>
		  body { margin:0; padding: 0; width:100%; height:100%;}
		  .map { position:absolute; top:0; bottom:0; width:100%; }
    </style>
    <link rel="stylesheet" href="https://unpkg.com/leaflet@1.0.3/dist/leaflet.css">
    <script src="https://unpkg.com/leaflet@1.0.3/dist/leaflet.js"></script>
    <script type="text/javascript" src="https://unpkg.com/leaflet.vectorgrid@1.2.0"></script>
    <script src="https://unpkg.com/proj4leaflet@1.0.2/lib/proj4-compressed.js"></script>
    <script src="https://unpkg.com/proj4leaflet@1.0.2/src/proj4leaflet.js"></script>
  </head>
  <body>
    <div id="map" class="map"></div>

	<script type="text/javascript">
      var crs = new L.Proj.CRS(
          'http://www.opengis.net/def/crs/EPSG/0/2056',
          '+proj=somerc +lat_0=46.9524055555556 +lon_0=7.43958333333333 +k_0=1 +x_0=2600000 +y_0=1200000 +ellps=bessel +units=m +no_defs +type=crs', {
          origin: [2420000.0, 1350000.0],
          bounds: L.bounds(
            L.Point(2420000.0, 326000.0),
            L.Point(3444000.0, 1350000.0)
          ),
          resolutions: [4000.0, 3750.0, 3500.0, 3250.0, 3000.0, 2750.0, 2500.0, 2250.0, 2000.0, 1750.0, 1500.0, 1250.0, 1000.0, 750.0, 650.0, 250.0, 100.0, 50.0, 20.0, 20.0, 10.0, 5.0, 2.5, 2.0, 1.5, 1.0, 0.5],
        }
      );

      var map = L.map('map', {
        center: [47, 10],
        zoom: 3,
        maxZoom: 24,
        crs: crs
      });

      L.tileLayer(
      "https://wmts.geo.admin.ch/1.0.0/ch.swisstopo.swissimage/default/current/2056/{z}/{x}/{y}.jpeg",
    ).addTo(map);

      L.vectorGrid.protobuf(
        "http://127.0.0.1:8000/collections/public.gg_gemeindegrenze_mv/tiles/2056_26/{z}/{x}/{y}",
      ).addTo(map)
    </script>

  </body>
</html>

```

I get something that looks ok 

<img width="878" src="https://github.com/developmentseed/tipg/assets/10407788/2d42d5dd-7c6d-43ef-af98-68f54cd8f74a">

I've tried with OpenLayer but without success 🤷

**Comment by p1d1d1** [comment]
Just FYI, last year we were able to run a service providing VT in CRS:2056 with t-rex https://github.com/t-rex-tileserver/t-rex.
It works with OL (see usergrid.html and usergrid.toml in the example folder).

PS: the example are based on resolutions available in wmts.geo.admin.ch (2056) which are non exactly those of SwissLV95CellSizes

**Reply by p1d1d1** [reply]
... and here https://github.com/p1d1d1/p1d1d1.github.io/tree/master/vt a quickly modified version of the 2 files. Using SwissLV95CellSizes as TMS for the VT and a WMS layer as background. It seems to work
